src/gtk/toga_gtk/widgets/table.py

```diff
@@ -4,53 +4,78 @@
             renderer = Gtk.CellRendererText()
             column = Gtk.TreeViewColumn(heading, renderer, text=i)
-            self.table.append_column(column)
+            self.treeview.append_column(column)
 
         self.native = Gtk.ScrolledWindow()
         self.native.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
-        self.native.add(self.table)
+        self.native.add(self.treeview)
         self.native.set_min_content_width(200)
         self.native.set_min_content_height(200)
         self.native.interface = self.interface
 
-    def refresh(self):
-        self.data.clear()
+    def _on_select(self, selection):
+        if hasattr(self.interface, "_on_select") and self.interface.on_select:
+            tree_model, tree_iter = selection.get_selected()
 
-        for row in self.interface.data.rows:
-            self.data.append(row.data)
+            if tree_iter:
+                tree_path = tree_model.get_path(tree_iter)
+                row = tree_path.get_indices()[0]
+            else:
+                row = None
+            self.interface.on_select(None, row=row)
 
-    def set_on_select(self, handler):
+    def _refresh(self):
+        """
+        Synchronize self.data with self.interface.data
+        """
 
-        for conn_id in self._connections:
-            # Disconnect all other on_select handlers, so that if you reassign
-            # the on_select, it doesn't trigger the old ones too.
-            self.selection.disconnect(conn_id)
-            self._connections.remove(conn_id)
+        # Temporarily disconnecting the TreeStore improves performance for large
+        # updates by deferring row rendering until the update is complete.
+        self.treeview.set_model(None) # temporarily disconnect the view
 
-        if handler is None:
-            return
+        self.store.clear()
+        for row in self.interface.data:
+            self.store.append(self._row_items(row))
 
-        def _handler(selection):
-            tree_model, tree_iter = selection.get_selected()
+        self.treeview.set_model(self.store)
 
-            if tree_iter:
-                tree_path = tree_model.get_path(tree_iter)
-                index = tree_path.get_indices()[0]
-                handler(self.table, row=index)
-            else:
-                handler(self.table, row=None)
+    # TODO: The interface should provide the row index instead of using this
+    # method to access a private property
+    def _row_index(self, row):
+        return self.interface.data._data.index(row)
+
+    # TODO: The interface should provide the row list items instead of using
+    # this method to access a private property
+    def _row_items(self, row):
+        return [getattr(row, attr) for attr in self.interface._accessors]
+
+    def change_source(self, source):
+        self._refresh()
+
+    def insert(self, index, item):
+        self.store.insert(index, self._row_items(item))
```

Comment:
Your test case worked! However, I found some [old docs for pygtk2] (http://www.pygtk.org/pygtk2tutorial/sec-TreeModelInterface.html) stating that not all versions of Gtk are consistent with this behavior. You can check if a row iters are persisent with `treestore.get_flags()`, which will include the `TREE_MODEL_ITERS_PERSIST` flag. I suppose my helpers could be used a fallback when that flag is absent, but there is another option altogether, call Gtk.TreeRowReference. Those are always persistent, though slightly less performant. Another [source](http://scentric.net/tutorial/sec-treemodel-rowref.html) states that both `Gtk.ListStore` and `Gtk.TreeStore` use persistent iters, but that `Gtk.TreeRowReference` is the "Right Way". 
Sidenote: It always irks me when a problem has two solutions and the more performant one is also the discouraged one :)

In the future, it may be worth making some benchmarks or something to determine which of the solutions on the table is most suitable (currently, these are my helper functions, persistent Gtk.TreeIter, and Gtk.TreeRowReference).

Anyway, I don't mind refactoring a little to use `Gtk.TreeRowReference` as the Node `_impl`. Although my research shows that using the iter returned from `insert` should [theoretically] be fine, something in me sees it as a risk of unexpected behavior.

Unless you have further opinion on this matter, I'll go with Gtk.TreeRowReference since it's purpose is clearly for keeping track of rows as the data changes. If we come up with something better in the future we can adapt.